src/providers/tenderly-simulation-provider.ts

```diff
@@ -0,0 +1,324 @@
+  private tenderlyBaseUrl: string;
+  private tenderlyUser: string;
+  private tenderlyProject: string;
+  private tenderlyAccessKey: string;
+  private v2PoolProvider: IV2PoolProvider;
+  private v3PoolProvider: IV3PoolProvider;
+
+  constructor(
+    tenderlyBaseUrl: string,
+    tenderlyUser: string,
+    tenderlyProject: string,
+    tenderlyAccessKey: string,
+    v2PoolProvider: IV2PoolProvider,
+    v3PoolProvider: IV3PoolProvider
+  ) {
+    this.tenderlyBaseUrl = tenderlyBaseUrl;
+    this.tenderlyUser = tenderlyUser;
+    this.tenderlyProject = tenderlyProject;
+    this.tenderlyAccessKey = tenderlyAccessKey;
+    this.v2PoolProvider = v2PoolProvider;
+    this.v3PoolProvider = v3PoolProvider;
+  }
+
+  public async simulateTransaction(
+    fromAddress: string,
+    swapRoute: SwapRoute,
+    l2GasData?: ArbitrumGasData | OptimismGasData
+  ): Promise<SwapRoute> {
+    const currencyIn = swapRoute.trade.inputAmount.currency;
+    const tokenIn = currencyIn.wrapped;
+    const chainId = tokenIn.chainId;
+    if ([ChainId.CELO, ChainId.CELO_ALFAJORES].includes(chainId)) {
+      const msg = 'Celo not supported by Tenderly!';
+      log.info(msg);
+      throw new Error(msg);
+    }
+
+    if (!swapRoute.methodParameters) {
+      throw new Error('No calldata provided to simulate transaction');
+    }
+    const { calldata } = swapRoute.methodParameters;
+    log.info(
+      {
+        calldata: swapRoute.methodParameters.calldata,
+        fromAddress: fromAddress,
+        chainId: chainId,
+        tokenInAddress: tokenIn.address,
+      },
+      'Simulating transaction via Tenderly'
+    );
+
+    const approve = {
+      network_id: chainId,
+      input: APPROVE_TOKEN_FOR_TRANSFER,
+      to: tokenIn.address,
+      value: '0',
+      from: fromAddress,
+      gasPrice: '0',
+      gas: 30000000,
+    };
+
+    const swap = {
+      network_id: chainId,
+      input: calldata,
+      to: SWAP_ROUTER_ADDRESS,
+      value: currencyIn.isNative ? swapRoute.methodParameters.value : '0',
+      from: fromAddress,
+      gasPrice: '0',
+      gas: 30000000,
+      type: 1,
+    };
+
+    const body = { simulations: [approve, swap] };
+    const opts = {
+      headers: {
+        'X-Access-Key': this.tenderlyAccessKey,
+      },
+    };
+    const url = TENDERLY_BATCH_SIMULATE_API(
+      this.tenderlyBaseUrl,
+      this.tenderlyUser,
+      this.tenderlyProject
+    );
+    const resp = (await axios.post<TenderlyResponse>(url, body, opts)).data;
+
+    // Validate tenderly response body
+    if (
+      !resp ||
+      resp.simulation_results.length < 2 ||
+      !resp.simulation_results[1].transaction ||
+      resp.simulation_results[1].transaction.error_message
+    ) {
+      const msg = `Failed to Simulate Via Tenderly!: ${resp.simulation_results[1].transaction.error_message}`;
+      log.info(
+        { err: resp.simulation_results[1].transaction.error_message },
+        msg
+      );
+      throw new Error(msg);
+    }
```

Comment:
Also commented this in the last review. Should definitely return simulationFailed = true for this one.